In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


In [2]:
import pandas as pd
churn_df = pd.read_csv("/content/gdrive/My Drive/ChurnData.csv")
churn_df.head()

tenure   age  address  income   ed  ...  loglong  logtoll  lninc  custcat  churn
0    11.0  33.0      7.0   136.0  5.0  ...    1.482    3.033  4.913      4.0    1.0
1    33.0  33.0     12.0    33.0  2.0  ...    2.246    3.240  3.497      1.0    1.0
2    23.0  30.0      9.0    30.0  1.0  ...    1.841    3.240  3.401      3.0    0.0
3    38.0  35.0      5.0    76.0  2.0  ...    1.800    3.807  4.331      4.0    0.0
4     7.0  35.0     14.0    80.0  2.0  ...    1.960    3.091  4.382      3.0    0.0

[5 rows x 28 columns]

In [0]:
churn_df = churn_df[['tenure', 'age', 'address', 'income', 'ed', 'employ', 'equip', 'callcard', 'wireless','churn']] #churn is out response variable

In [0]:
churn_df['churn'] = churn_df['churn'].astype('int')

In [5]:
churn_df.head() #only display five or certain number of item from the data base. 

tenure   age  address  income   ed  employ  equip  callcard  wireless  churn
0    11.0  33.0      7.0   136.0  5.0     5.0    0.0       1.0       1.0      1
1    33.0  33.0     12.0    33.0  2.0     0.0    0.0       0.0       0.0      1
2    23.0  30.0      9.0    30.0  1.0     2.0    0.0       0.0       0.0      0
3    38.0  35.0      5.0    76.0  2.0    10.0    1.0       1.0       1.0      0
4     7.0  35.0     14.0    80.0  2.0    15.0    0.0       1.0       0.0      0

In [6]:
churn_df

tenure   age  address  income  ...  equip  callcard  wireless  churn
0      11.0  33.0      7.0   136.0  ...    0.0       1.0       1.0      1
1      33.0  33.0     12.0    33.0  ...    0.0       0.0       0.0      1
2      23.0  30.0      9.0    30.0  ...    0.0       0.0       0.0      0
3      38.0  35.0      5.0    76.0  ...    1.0       1.0       1.0      0
4       7.0  35.0     14.0    80.0  ...    0.0       1.0       0.0      0
..      ...   ...      ...     ...  ...    ...       ...       ...    ...
195    55.0  44.0     24.0    83.0  ...    0.0       1.0       0.0      0
196    34.0  23.0      3.0    24.0  ...    0.0       1.0       0.0      0
197     6.0  32.0     10.0    47.0  ...    0.0       1.0       0.0      0
198    24.0  30.0      0.0    25.0  ...    0.0       1.0       1.0      1
199    61.0  50.0     16.0   190.0  ...    1.0       1.0       1.0      0

[200 rows x 10 columns]

In [7]:
churn_df.shape

(200, 10)

In [8]:
import numpy as np

X = np.asarray(churn_df[['tenure', 'age', 'address', 'income', 'ed', 'employ', 'equip']]) # Now I want to use the 7 indenpendent variables to predict the likelihood of customer's churn"
X[0:5]

array([[ 11.,  33.,   7., 136.,   5.,   5.,   0.],
       [ 33.,  33.,  12.,  33.,   2.,   0.,   0.],
       [ 23.,  30.,   9.,  30.,   1.,   2.,   0.],
       [ 38.,  35.,   5.,  76.,   2.,  10.,   1.],
       [  7.,  35.,  14.,  80.,   2.,  15.,   0.]])

In [9]:
y = np.asarray(churn_df['churn'])
y[0:5]

array([1, 1, 0, 0, 0])

In [11]:
# X, y is now the data set we will process in this practice
#There is an issue in here... 
# The units are different... in X. --> Normalizing the data..(When you hae more than one independent varaible, you should do this before you train the data) 

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)
X[0:5]


array([[-1.13518441, -0.62595491, -0.4588971 ,  0.4751423 ,  1.6961288 ,
        -0.58477841, -0.85972695],
       [-0.11604313, -0.62595491,  0.03454064, -0.32886061, -0.6433592 ,
        -1.14437497, -0.85972695],
       [-0.57928917, -0.85594447, -0.261522  , -0.35227817, -1.42318853,
        -0.92053635, -0.85972695],
       [ 0.11557989, -0.47262854, -0.65627219,  0.00679109, -0.6433592 ,
        -0.02518185,  1.16316   ],
       [-1.32048283, -0.47262854,  0.23191574,  0.03801451, -0.6433592 ,
         0.53441472, -0.85972695]])

In [0]:
#Train the model with the data above
#We will use the loglinear, because the churn variable(response variable) has only two class (1 (left for the other company), 0 (loyal customer))

#Split data --> Training data, Test data... --> Cross-validation (in practically reduce the Bias-Variance tradeoff)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=4) #random_state is optional, that indicates the initial random seed. 


In [16]:
X_test[0:5]

array([[ 0.53250132, -0.08931261,  0.42929083,  0.40488962,  0.91629947,
         0.19865678,  1.16316   ],
       [ 0.81044894, -0.62595491, -1.05102238, -0.11810256, -1.42318853,
        -0.4728591 , -0.85972695],
       [-0.25501694, -1.00927084, -1.05102238, -0.4303367 , -1.42318853,
        -0.24902047, -0.85972695],
       [ 0.76412434, -0.1659758 , -0.55758464,  0.34244279,  0.13647013,
         0.87017266, -0.85972695],
       [-0.11604313, -0.62595491,  0.03454064, -0.32886061, -0.6433592 ,
        -1.14437497, -0.85972695]])

In [17]:
X_test.shape

(40, 7)

In [0]:
#define sigmoid function...
def sigmoid(z):
    return 1/(1+np.exp(-z))

In [23]:
theta = [0, 0, 0, 0, 0, 0, 0] # simply setting the initial theta, this number wull be changed by Gradient Descent method
# theta is the parameter we will seek for finding by Gradient Descent method --> "You need to read more information of Loglinear regression to mean theta"
z = np.dot(X, theta) 
h = sigmoid(z) #our training purpose would be how this h can well represent our 'Churn' data probability. 

# You need to understand this by the next week. why Blake has used this form| 

array([0.5, 0.5, 0.5, 0.5, 0.5])

In [0]:
def loss(h,y):
    return ((-y*np.log(h) -(1-y)*np.log(1-h)).mean()) #why I have use the mean() function to define the loss function

#Need to understand why this loss function is set for loglinear regression, this has not been explained in the textbook. 
#This would be the first and last time to see the loss function in machine learning, because we will use the Python library in your future. 

In [25]:
loss(h, y) # loss function, to minimize... our main job is to reduce loss as much as we can..... almost 0... 

0.6931471805599452

the value above is the outcome of loss function. We need to minimize it, how? to move down following the best gradient at this point

In [0]:
gradient = np.dot(X.T, (h-y))/y.shape[0] #This is the gradient at this point (where the loss value is 0.6931471805599452)
# derivatives of loss function on theta. 

Trying to understand why such gradient is made in Blake's programming. --> You need to study it

In [27]:
gradient

array([ 0.17100496,  0.13054591,  0.11827703,  0.04119695, -0.09806354,
        0.15335744, -0.12491327])

Using the gradient values(i.e., 0.17100496,  0.13054591,  0.11827703,  0.04119695, -0.09806354,
        0.15335744, -0.12491327], you will use the Newton method to find the next theta. 

This process from [23] to [27] repeating until the loss is getting to small and similar with the previous step. --> Gradient Descent Method (machine lenaring on numerical approach)

In [0]:
#prediction probability
def predict_probs(X, theta):
    return sigmoid(np.dot(X, theta)) #This theta should be the final best theta. 

def predict(X, theta, threshold = 0.5):
    return predict_probs(X, theta) >= threshold 

In [29]:
predict(X, theta) # at last, if you have a new data (or test data), call the final best model , not the training data, will return the probability of the new data item. 

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,